### DISCUSSION TITLE
Extension with kotlinx.coroutines

### DISCUSSION THREAD

**Discussion by pank-su** [Kotlin/kotlinx-datetime#338] [Ideas]
Sometimes in the apps, you need to know when time is changed, for example: deadlines in TODO. It was very helpfull, if kotlinx.datetime has flow extensions.

For example

```kotlin
Clock.System.now().asDateFlow() // emitted every date (day) change
```

**Comment by dkhalanskyjb** [comment]
Is your use case an Android project? If so, I don't think the proposed API would work: Android reserves the right to suspend your process when you're not using it, and you have to rely on Android's `WorkManager` to reliably schedule tasks. `asDateFlow` (no matter how we implement it) would only reliably emit dates if the app is open at the moment the emission is due.

**Reply by pank-su** [reply]
Yes, you right, but kotlinx corutines and datetime is multiplatform, and on not on every platform you can use workmanager^_^. 

Secondly, I think, is not bad then you need react on every interval time, for example simple timer or stopwatch, I know, you can just use delay and duration, but in the future it will be cool than you update UI by one line code.

But it's just my crazy idea, I think is not bad and not dificiult to release this. If I will have many more use cases to this future I write here.

Thank you to your reaction👍.

**Reply by dkhalanskyjb** [reply]
Let's focus on specific use cases. We don't want to add a function that no one will actually use or one that looks like it's going to solve a use case, but in fact will silently do something different.

* Multiplatform way to run some code when the date changes. As I mentioned, this won't work on Android, so a multiplatform library developer or application will have to use `expect`/`actual` anyway to provide proper Android support, so there's no gain.
* You are talking about stopwatches and timers, but they have nothing to do with `kotlinx-datetime`: a timer/stopwatch starts counting seconds from the moment you press a button, so knowing when the second (or minute, or hour, or day) changed on your wall clock is meaningless.
* Anything else?

**Reply by kevincianfarini** [reply]
If you don't require your job scheduling to live outside of your apps process death, then you might want to take a look at [Cadiologist](https://github.com/kevincianfarini/cardiologist) -- a library that I maintain -- for scheduling recurring jobs built on top of kotlinx-datetime and kotlinx-coroutines. 

```kt
val tz = TimeZone.of("America/New_York")
Clock.System.schedulePulse(timeZone = tz, atSecond = 0).beat { scheduled, occurred ->
    println("A pulse was scheduled in $tz for ${scheduled.toLocalDateTime(tz)} and occurred at ${occurred.toLocalDateTime(tz)}.")
}
```

```
A pulse was scheduled in America/New_York for 2025-01-11T08:41 and occurred at 2025-01-11T08:41:00.005227606
A pulse was scheduled in America/New_York for 2025-01-11T08:42 and occurred at 2025-01-11T08:42:00.001463582
A pulse was scheduled in America/New_York for 2025-01-11T08:43 and occurred at 2025-01-11T08:43:00.001028731
```
